### ISSUE TITLE
Make HTTP transport easy without tightly coupling to OT

### ISSUE THREAD

**Issue by jmacd** [opentracing/opentracing.github.io#98]
As discussed at the OT weekly meeting, there is a problem caused by opentracing-go's use of url.QueryEscape while setting keys in the HTTPHeaderTextMapCarrier. The problem is that other languages need to be aware of this in order to join, and how to encode keys into HTTP is not part of the OT spec (and it shouldn't be).

It is important to keep encoding specifications out of the OT spec, because OT is explicitly transport-agnostic, but there is a competing desire to provide short, easy-to-use APIs to allow make calls to Inject/Join short and readable at call sites. Thus, opentracing-go provides HTTPHeaderTextMapCarrier which makes Injecting to / Joining from an HTTP Request object a one-liner.

The discussion was how to decouple the ease-of-use API from the strictly-specified, transport-agnostic OT interface. General consensus is that we should provide a library of language-specific, not-part-of-the-OT-spec-proper helpers, possibly in a subdirectory of each language's OT repository, to support common transports in that language using whatever "normal" encoding is for that class of transport.

HTTPHeaderTextMapCarrier would go into such a helper directory, and similar helpers would be created in other languages / repos. These should be documented to use a specific encoding with one header per key:value, for example. These can be standards without being part of the OT spec.

**Comment by jmacd** [comment]
@bensigelman @yurishkuro @bg451 I think I've represented this issue. (Thanks to Ben for helping me understand this issue.)

**Comment by yurishkuro** [comment]
Is it fair to say that helper functions like `SpanFromContext` should also go into this "util" library?

fwiw, the way Go package structure works, this can be simply a directory in the opentracing-go repo. Similarly, if we create a (thread local) context propagation helpers in Java, it can be just a submodule released as an extra jar.

**Comment by jmacd** [comment]
Yes, I'd say so.

**Comment by bg451** [comment]
Overall I'm in favor of the proposal, +1 from me.

**Comment by bensigelman** [comment]
As @jmacd knows, I think this is a good compromise...

**Comment by yurishkuro** [comment]
@bensigelman I was thinking more about this and allowing baggage keys to be utf8, and it occurred to me that even if we move http encoding concerns into a sub-package, we still have a similar issue with keys returned in the TEXT_MAP carrier. Namely that by moving TEXT_MAP => HTTP conversion into _shared_ libraries we take away tracer's ability to control what gets put into HTTP headers, and consequently push http encoding concerns back into TEXT_MAP carrier specification.

I think the only clean way to completely divorce these two is by making something like HTTP_HEADER carrier type a mandatory format for OT tracers. That puts the tracer back in full control of what happens with HTTP. We resisted introducing protocol-specific standard formats previously, out of concern that it might lead to proliferation of standard formats, e.g. for GRPC, Thrift, ArrrrPC, etc. However, I think with most other formats the situation is much simpler in that they don't have these dumb lax rules about headers the way HTTP has. I.e. most binary formats would preserve any key/value in the headers without loss, therefore a TEXT_MAP format is perfectly suitable for those.

**Comment by bensigelman** [comment]
@yurishkuro you are a smart guy so I think I'm probably missing something, but...

While I see how the Tracer doesn't have as much control over the HTTP layer if a helper method is used instead, I *don't* see why the encoding concerns end up in TEXT_MAP. Wouldn't the contract for the HTTP impl of TEXT_MAP specify precisely how keys and values are transformed to be HTTP-header-compliant? And couldn't those transformations be safely reversed on the other side to yield the free-form utf8 key:value pairs?

**Comment by yurishkuro** [comment]
ok, maybe it's possible to come up with this bullet-proof encoding, and we just haven't done it in any library. Note that utf8 is not so much of a concern there, because in the extreme case a text map carrier can be populated with exactly one key, and the value will encode both trace ids and baggage in some way. Let's say the key is `Lightstep-Trace` - how can we encode it into http header name with a guarantee that it will come up from decode exactly the same?
  * we can use base64, it will work, but the header name will look like garbage on the wire
  * we can use a stronger form of url encoding, e.g. forcing all upper-case letters to %xx code and only allowing lower-case letters: `%43ightstep-%64race` on the wire

I think any encoding scheme will look bad on the wire _unless_ the tracer knows the internal details of how the shared library does encoding and picks the key name that will pass that encoding unchanged. That's clearly pushing http encoding concern into TEXT_MAP handling.

Another nasty aspect the shared library needs to deal with is filtering out non-tracing headers during decoding. Say we picked base64 approach above. The decode method will get all http headers in the request, it will have no way of knowing which ones are relevant, and will have to try base64 decode on all of them... Or play the prefix game, with the prefix shared by _all_ traces (like OT_TRACE_).

So to summarize, I concede that you're technically correct (the best kind of correct) that shared encoding can deal with the issue, but the solution wouldn't be anywhere as clean as it could be if the tracer was given full control over http encoding. And if some tracers don't want to deal with that, we can still have that shared library they can fall back to when they see HTTP_HEADER format. From the end user /  instrumentation code point of view I do not see a lot of differences, they either use HTTP_HEADER format, or they are forced to use TEXT_MAP format in conjunction with the shared library. The former sounds a lot more flexible to me.

**Comment by bensigelman** [comment]
@yurishkuro:

> I think any encoding scheme will look bad on the wire unless the tracer knows the internal details of how the shared library does encoding and picks the key name that will pass that encoding unchanged. That's clearly pushing http encoding concern into TEXT_MAP handling.

This is the essence of the issue, I think. So, just to be clear: Yuri, you are voting for an HTTP_HEADER format, correct?

**Comment by yurishkuro** [comment]
Yes, that would be my preference. Instead of this
```go
tracer.Inject(
                span,
                opentracing.TextMap,
                opentracing.HTTPHeaderTextMapCarrier(httpReq.Header))
```

the instrumentation would do this:

```go
tracer.Inject(span, opentracing.HTTPHeader, httpReq.Header)
```

This change has the best chance of being backwards compatible with existing prod installations.

**Comment by bensigelman** [comment]
Sorry for the delay, I have been thinking about this and hoping for a revelation. (I didn't have one)

In Go this seems like a worthwhile compromise. In languages/platforms without a single canonical type for http headers, it's less clear... we muddy the waters with two sets of interfaces — one for TextMap, one for HTTPHeaderMap — which have the same type signatures but just different names and comments. And not so that tracers are more correct, but so that tracers can guarantee encoding for HTTP specifically.

An alternative would be to take advantage of the extensibility of "carrier" and use something like `not_opentracing.HTTPHeader` for the carrier type argument. If there are indeed many organizations who need this sort of explicit control over the HTTP headers, then `not_opentracing.HTTPHeader` could be promoted to `opentracing.HTTPHeader`. Needless to say, this proposal is not as friendly from an interop standpoint.

Thoughts??

**Comment by yurishkuro** [comment]
I don't like resorting to `non_opentracing`, since by definition it makes the instrumentation code not compatible with all compliant OT tracers.

> In languages/platforms without a single canonical type for http headers, it's less clear... 

Currently, in Python we have to do this:
```python
    carrier = {}
    tracer.inject(span, Format.TEXT_MAP, carrier)
    for key, value in carrier.iteritems() {
        req.headers[key] = urllib.quote(value)
    }
```
After introducing `HTTP_HEADER` format:
```python
    tracer.inject(span, Format.HTTP_HEADER, req.headers)
```
which will work out of the box for the majority of http frameworks because they usually make headers implement a `dict` api.

Java case is less clear, but only because we haven't actually defined any standard formats. If we did, we'd have the same leak of http encoding into TEXT_MAP carrier definition as we have in Go:

```
	// NOTE: Since HTTP headers are a particularly important use case for the
	// TextMap carrier, `key` parameters identify their respective values in a
        // case-insensitive manner.
```

**Comment by yurishkuro** [comment]
I have a glitch in my proposal w.r.t. Go, will post a clarified version shortly

**Comment by yurishkuro** [comment]
So here's updated proposal.

### Objectives
  1. Remove all encoding restrictions from `TEXT_MAP` format (e.g. case insensitivity and other http quirks)
  1. Push the responsibility of http encoding from users to tracer

### Proposal
Introduce a new standard format `HTTP_HEADERS`:
  * Its semantics should be identical to `TEXT_MAP`, with the only distinction that the keys and values are safe for HTTP protocol
  * It should use the same carrier interfaces as `TEXT_MAP` format
    * Go: `TextMapWriter` and `TextMapReader`, but with understanding that keys/values are safe for http
    * Python: `dict` that is safe to use in http
    * Java: once we define (I hope) `TextMapWriter` and `TextMapReader`, the `HTTP_HEADERS` format will require `HTPHeadersWriter` and `HTTPHeadersReader` extension interfaces since the carrier is identified by type, not by an enum

Note that there is a change in Go from my earlier proposal, to remove OT dependency on standard HTTPHeader type in case people find themselves using different http frameworks. The Go example will look similar to today:
```go
tracer.Inject(span, opentracing.HTTPHeader, opentracing.HTTPHeaderCarrier(httpReq.Header))
```
where the `HTTPHeaderCarrier` helper class extends standard `HTTPHeader` type with `TextMapReader` and `TextMapWriter` methods as it does today, but *does not* apply url encoding, leaving it to the tracer.

**Comment by bensigelman** [comment]
@yurishkuro, FWIW I like the move to interfaces (from the concrete `http.Header` type) in this most recent proposal.

I have thought about this enough to realize that I'm not going to particularly like any of the available options. :) I will never like the aesthetics of having OT depend so directly on HTTP concepts/types. I will also never like the ramifications for header keys and values if OT *doesn't* depend so directly on HTTP! What a bummer.

If I consider everything and make a judgement call, I'd side with Yuri's proposal. The tie-breaking issue for me is the fact that HTTP header keys often need to be known ahead of time to allow for passthrough (not to mention CORS, etc) in various places; without giving Tracers specific control over HTTP wire details, it will be difficult to explain the exact spelling of HTTP header keys as they're sent over the wire.

**Comment by bensigelman** [comment]
Java: https://github.com/opentracing/opentracing-java/tree/master/opentracing-api/src/main/java/io/opentracing/propagation or https://github.com/opentracing/opentracing-java/pull/33... either way we're done.

Javascript: https://github.com/opentracing/opentracing-javascript/pull/41

Go: https://github.com/opentracing/opentracing-go/pull/102

Python: https://github.com/opentracing/opentracing-python/pull/27

**Comment by bensigelman** [comment]
Phew, all done with ^^^. Closing!
